Broker Relist

eriknelson opened the pull request:
Initial fields for #1086

pmorie commented on pkg/apis/servicecatalog/types.go:
```diff
@@ -60,8 +60,29 @@ type ServiceBrokerSpec struct {
 	// +optional
 	CABundle []byte
+
+	// RelistBehavior specifies the type of relist behavior the catalog should exhibit
```
We need to add validations for these fields.  See pkg/apis/servicecatalog/validation/broker.go

pmorie commented on pkg/apis/servicecatalog/v1alpha1/types.go:
```diff
@@ -60,8 +60,29 @@ type ServiceBrokerSpec struct {
+
+	// RelistBehavior specifies the type of relist behavior the catalog should exhibit
+	// when relisting ServiceClasses available from a broker
+	RelistBehavior ServiceBrokerRelistBehavior
```
fields in versioned APIs need to have json tags associated with them.  For this field, it would be:

```go
RelistBehavior ServiceBrokerRelistBehavior `json:"relistBehavior"`
```

You'll need to add these to the fields in this file.

pmorie commented on pkg/controller/controller_broker.go:
```diff
@@ -113,8 +113,8 @@ const (
 // the controller's broker relist interval has not elapsed since the broker's
-// ready condition became true.
-func shouldReconcileServiceBroker(broker *v1alpha1.ServiceBroker, now time.Time, relistInterval time.Duration) bool {
+// ready condition became true, or if the broker's RelistBehavior is set to Manual
```
missing period

pmorie commented on pkg/controller/controller_broker.go:
```diff
@@ -113,8 +113,8 @@ const (
-// ready condition became true.
-func shouldReconcileServiceBroker(broker *v1alpha1.ServiceBroker, now time.Time, relistInterval time.Duration) bool {
+// ready condition became true, or if the broker's RelistBehavior is set to Manual
+func shouldReconcileServiceBroker(broker *v1alpha1.ServiceBroker, now time.Time, relistInterval metav1.Duration) bool {
```
Technically you do not need to pass in the `relistInterval` parameter here, since you're now reading it from the broker.

pmorie replied:
Oh, I didn't realize you were still reading it as a parameter.  You already have the broker here - you can read the value in this method and delete the `relistInterval` parameter.  There needs to ve a validation that ensures that the duration field is set when the mode is 'duration'.

pmorie commented on pkg/controller/controller_broker.go:
```diff
@@ -168,14 +185,16 @@ func (c *controller) reconcileServiceBrokerKey(key string) error {
 // processed and should be resubmitted at a later time.
 func (c *controller) reconcileServiceBroker(broker *v1alpha1.ServiceBroker) error {
 	glog.V(4).Infof("Processing ServiceBroker %v", broker.Name)
+	if broker.Spec.RelistDuration == nil {
```
I think this TODO is fixed :)

You should be defensive in programming, but some assumptions that you have validated for already are okay.  So for instance, you should ensure that you don't dereference a nil pointer for an optional field.  However, if that optional field must be set based on a validation, a state where that validation fails in code should be handled like an error.

pmorie commented on pkg/controller/controller_broker.go:
```diff
@@ -168,14 +185,16 @@ func (c *controller) reconcileServiceBrokerKey(key string) error {
-			"Not processing ServiceBroker %v because relist interval has not elapsed since the broker became ready",
-			broker.Name,
-		)
+	if !shouldReconcileServiceBroker(broker, time.Now(), *broker.Spec.RelistDuration) {
```
Your TODO appears to be handled now in `shouldReconcileServiceBroker`.

pmorie commented on pkg/controller/controller_broker_test.go:
```diff
@@ -553,6 +554,8 @@ func testReconcileServiceBrokerWithAuth(t *testing.T, authInfo *v1alpha1.Service
 }
 
+// ERIK:TODO: Write up another test here to confirm an error is bubbled when
+// RelistDuration is nil, similar to this?
```
RelistDuration being nil should be handled in `shouldReconcileBroker`.

pmorie commented on pkg/storage/tpr/storage_interface_test.go:
```diff
@@ -140,7 +141,9 @@ func TestCreateWithNoNamespace(t *testing.T) {
 		ObjectMeta: metav1.ObjectMeta{Name: name},
 		Spec: sc.ServiceBrokerSpec{
-			URL: "http://my-awesome-broker.io",
+			URL:            "http://my-awesome-broker.io",
```
I am not qualified to review TPR stuff.  @arschles @nilebox

pmorie commented on pkg/apis/servicecatalog/testing/fuzzer.go:
```diff
@@ -192,6 +193,11 @@ func FuzzerFor(t *testing.T, version schema.GroupVersion, src rand.Source) *fuzz
 			// Set the bytes field on the RawExtension
 			r.Raw = bytes
 		},
+		func(bs *servicecatalog.ServiceBrokerSpec, c fuzz.Continue) {
```
I'm surprised that the default fuzzer didn't work for this field.  Do you try it?

eriknelson replied:
Think I did, but I'll pull it and try again.

pmorie commented on pkg/apis/servicecatalog/v1alpha1/defaults.go:
```diff
@@ -18,13 +18,25 @@ package v1alpha1
+		spec.RelistBehavior = ServiceBrokerRelistBehaviorDuration
+	}
+
+	if spec.RelistDuration == nil {
```
You should only default relist duration if the relist behavior is "Duration".  It should not be defaulted for manual behavior.

pmorie commented on pkg/apis/servicecatalog/v1alpha1/defaults_test.go:
```diff
@@ -68,6 +68,22 @@ func roundTrip(t *testing.T, obj runtime.Object) runtime.Object {
 	return obj3
 }
 
+func TestSetDefaultServiceBroker(t *testing.T) {
```
Tests here need to change according to the logic as I commented above

pmorie replied:
There should be test cases for:

1.  Neither duration or behavior set
2.  Behavior set to manual
3.  Behavior set to duration but no duration provided

pmorie replied:
Still need test cases for these defaults.  You need to ensure that after an object starting in each of the states in the above comment is defaults, that the object is in the expected state.

pmorie commented on pkg/apis/servicecatalog/v1alpha1/types.go:
```diff
@@ -60,8 +60,29 @@ type ServiceBrokerSpec struct {
+
+	// RelistBehavior specifies the type of relist behavior the catalog should exhibit
+	// when relisting ServiceClasses available from a broker
+	RelistBehavior ServiceBrokerRelistBehavior `json:"relistBehavior"`
```
newline separate these to keep it consistent with the rest of the file, same for the other types.go

pmorie commented on pkg/apis/servicecatalog/validation/broker.go:
```diff
@@ -99,6 +99,12 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
 			)
 		}
 
+		if spec.RelistBehavior == sc.ServiceBrokerRelistBehaviorDuration && spec.RelistDuration == nil {
```
Question: do you think it is valid for duration to be set if the behavior is manual?  It might be.  If you switch from duration to manual temporarily, for example, you might not expect to have to unset the duration.  Consider what the 'least surprise' is here.

pmorie replied:
Also, I don't see unit tests for the validations

eriknelson replied:
> do you think it is valid for duration to be set if the behavior is manual? It might be. If you switch from duration to manual temporarily, for example, you might not expect to have to unset the duration.

Agreed. I would be surprised if I had to explicitly unset duration, and instead would expect the duration to simply be ignored.

pmorie commented on pkg/controller/controller_broker_test.go:
```diff
@@ -54,77 +54,82 @@ func TestShouldReconcileServiceBroker(t *testing.T) {
-	// internal: the time that has elapsed since now
 	// reconcile: whether or not the reconciler should run, the return of
 	// shouldReconcileServiceBroker
+	// ERIK:TODO: Test new Relist codepaths
```
Looks like this TODO still needs to be done, accurate?

eriknelson replied:
Have a local commit with a bunch of error handling I wanted to run by you. It tests the RelistDuration == nil path, but I want to make sure I hit the rest of them.

pmorie commented on pkg/apis/servicecatalog/types.go:
```diff
@@ -60,8 +60,31 @@ type ServiceBrokerSpec struct {
+	// when relisting ServiceClasses available from a broker
+	RelistBehavior ServiceBrokerRelistBehavior
+
+	//RelistDuration is the frequency by which a controller will relist the broker
```
There should be spaces after the `//` and all godoc comments for fields should end in periods.

pmorie commented on pkg/apis/servicecatalog/types.go:
```diff
@@ -60,8 +60,31 @@ type ServiceBrokerSpec struct {
+type ServiceBrokerRelistBehavior string
+
+const (
+	// ServiceBrokerRelistBehaviorDuration will allow the RelistDuration frequency to be configured
```
`// ServiceBrokerRelistBehaviorDuration indicates that the broker will be relisted automatically after the specified duration has passed.`

pmorie commented on pkg/apis/servicecatalog/types.go:
```diff
@@ -60,8 +60,31 @@ type ServiceBrokerSpec struct {
+const (
+	// ServiceBrokerRelistBehaviorDuration will allow the RelistDuration frequency to be configured
+	ServiceBrokerRelistBehaviorDuration ServiceBrokerRelistBehavior = "Duration"
+	// ServiceBrokerRelistBehaviorManual will configure the catalog to only perform future
```
`// ServiceBrokerRelistBehaviorManual indicates that the broker is only relisted when the spec of the broker changes.`

pmorie commented on pkg/apis/servicecatalog/validation/broker.go:
```diff
@@ -99,6 +99,12 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
 			)
 		}
 
```
Need a validation here to ensure that the relist behavior is set.

pmorie replied:
Still need to update the unit tests for this validation.

pmorie commented on pkg/controller/controller_broker.go:
```diff
@@ -133,10 +133,36 @@ func shouldReconcileServiceBroker(broker *v1alpha1.ServiceBroker, now time.Time,
+				}
+
+				if broker.Spec.RelistDuration == nil {
+					glog.V(10).Infof(
```
This should be logged at Error

jboyd01 commented on pkg/apis/servicecatalog/validation/broker_test.go:
```diff
@@ -208,12 +229,56 @@ func TestValidateServiceBroker(t *testing.T) {
+			valid: true,
+		},
+		{
+			name: "valid broker - manual behavior with behavior",
```
nit: name should be manual behavior with **duration**   instead of manual behavior with beavior

staebler commented on pkg/apis/servicecatalog/validation/broker.go:
```diff
@@ -98,13 +98,25 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
 
+	if "" == spec.RelistBehavior {
+		allErrs = append(allErrs,
+			field.Required(fldPath.Child("relistDuration"),
```
The path here should be `relistBehavior` and the comment should indicate that the broker must have a relistBehavior.

staebler commented on pkg/apis/servicecatalog/validation/broker.go:
```diff
@@ -98,13 +98,25 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
+	if spec.RelistBehavior == sc.ServiceBrokerRelistBehaviorDuration && spec.RelistDuration == nil {
+		allErrs = append(
+			allErrs,
+			field.Required(fldPath.Child("relistDuration"), "relistDuration must be set if behavior is set to Duration"),
```
This should probably be "relistDuration must be set if **relistBehavior** is set to Duration", so that it is clear which field is of importance.

pmorie replied:
Looks like this still needs to be done

staebler commented on pkg/apis/servicecatalog/validation/broker.go:
```diff
@@ -98,13 +98,25 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
+			field.Required(fldPath.Child("relistDuration"), "relistDuration must be set if behavior is set to Duration"),
+		)
+	}
+
```
Do we want to enforce that `relistRequests` be non-negative.

eriknelson replied:
I had a similar though. Should the field be a `uint`? I that normally acceptable?

staebler replied:
`uint` is discouraged. All you need to do is add a validation check to ensure it is non-negative. And indicate that restriction in the comment.

eriknelson replied:
@staebler pushed an update to the comments indicating as such. Validation asserts greater than zero.

staebler commented on pkg/apis/servicecatalog/v1alpha1/defaults_test.go:
```diff
@@ -68,6 +70,64 @@ func roundTrip(t *testing.T, obj runtime.Object) runtime.Object {
+			)
+		}
+
+		if tc.duration == nil && actualSpec.RelistDuration == nil {
```
This check does not verify that the actual RelistDuration was nil for the "behavior set to manual" test.

eriknelson replied:
There is an open question as to whether or not RelistDuration MUST BE nil given a RelistBehavior of Manual. Is this something that needs to be asserted? CC @pmorie

nilebox replied:
@eriknelson yes, it should be enforced to nil (and checked at the validation stage). similar to union types, where only one branch can be initialized at a time.

staebler commented on pkg/apis/servicecatalog/validation/broker.go:
```diff
@@ -98,13 +98,25 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
+				"brokers must have a relistDuration"))
+	}
+
+	if spec.RelistBehavior == sc.ServiceBrokerRelistBehaviorDuration && spec.RelistDuration == nil {
```
We need to validate that the RelistDuration is above a reasonable minimum, at least above 0.

staebler commented on pkg/apis/servicecatalog/validation/broker_test.go:
```diff
@@ -208,12 +229,56 @@ func TestValidateServiceBroker(t *testing.T) {
+				},
+			},
+			valid: false,
+		},
```
We also need tests to verify the behavior when RelistBehavior is invalid and empty.

staebler commented on pkg/controller/controller_broker.go:
```diff
@@ -131,12 +131,39 @@ func shouldReconcileServiceBroker(broker *v1alpha1.ServiceBroker, now time.Time,
+					return false
+				}
+
+				if broker.Spec.RelistDuration == nil {
```
Is this check necessary given that we have already performed validation on the Broker Spec to ensure that the RelistDuration is not nil when the RelistBehavior is not Manual?

eriknelson replied:
I've been back and forth on this. Are folks generally comfortable with trusting and dereffing a pointer value given a validation?

pmorie replied:
You should always program defensively, IMO.

staebler commented on pkg/apis/servicecatalog/types.go:
```diff
@@ -60,8 +60,31 @@ type ServiceBrokerSpec struct {
+	// when the RelistBehavior is set to ServiceBrokerRelistBehaviorDuration
+	RelistDuration *metav1.Duration
+
+	// RelistRequests is a strictly increasing integer counter that can be manually incremented
```
If we are claiming that this is a strictly increasing integer, then we should add a check in the ValidateUpdate function of the brokerRESTStrategy that validates that the new value is not less than the old value.

staebler commented on pkg/controller/controller_broker_test.go:
```diff
@@ -133,12 +138,33 @@ func TestShouldReconcileServiceBroker(t *testing.T) {
 			reconcile: true,
 		},
+		{
+			name: "ready, duration behavior, nil duration",
```
We shouldn't need to test this. This should be caught by the Broker validator.

eriknelson replied:
If it's being asserted in `shouldReconcileBroker`,  I think it's worth keeping as a test case. Happy to remove if folks feel strongly otherwise.

pmorie replied:
Keep it

staebler commented on pkg/controller/controller_broker_test.go:
```diff
@@ -553,6 +585,7 @@ func testReconcileServiceBrokerWithAuth(t *testing.T, authInfo *v1alpha1.Service
 	}
 }
 
+// RelistDuration is nil, similar to this?
```
Is this a stray comment that got placed here accidentally?

eriknelson commented:
@staebler responded to feedback, thank you!

pmorie commented on pkg/apis/servicecatalog/validation/broker.go:
```diff
@@ -98,13 +99,50 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
+	if "" == spec.RelistBehavior {
+		allErrs = append(allErrs,
+			field.Required(fldPath.Child("relistBehavior"),
+				"brokers must have a relistBehavior"))
```
"relist behavior is required"

pmorie commented on pkg/apis/servicecatalog/validation/broker.go:
```diff
@@ -98,13 +99,50 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
+	isValidRelistBehavior := spec.RelistBehavior == "Duration" ||
+		spec.RelistBehavior == "Manual"
+	if !isValidRelistBehavior {
+		errMsg := fmt.Sprintf("unknown relistBehavior found: %v", spec.RelistBehavior)
```
"relist behavior must be "Manual" or "Duration""

staebler replied:
Or use field.NotSupported which takes a []string of valid values.

staebler replied:
Or use field.NotSupported which takes a []string of valid values.

pmorie commented on pkg/apis/servicecatalog/validation/broker.go:
```diff
@@ -98,13 +99,50 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
+				"brokers must have a relistBehavior"))
+	}
+
+	isValidRelistBehavior := spec.RelistBehavior == "Duration" ||
```
use the constants from API types

pmorie commented on pkg/apis/servicecatalog/validation/broker.go:
```diff
@@ -98,13 +99,50 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
+				field.Required(fldPath.Child("relistDuration"), "relistDuration must be set if behavior is set to Duration"),
+			)
+		}
+		if spec.RelistRequests < 0 {
```
I think you always want to apply this validation, even if the behavior is manual

pmorie commented on pkg/apis/servicecatalog/validation/broker.go:
```diff
@@ -98,13 +99,50 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
+	}
+
+	if spec.RelistBehavior == sc.ServiceBrokerRelistBehaviorDuration {
+		if spec.RelistDuration == nil {
```
looks like this is a duplicate of :124, and looks like this can be removed now, see comment about non-negative validation

pmorie commented on pkg/apis/servicecatalog/validation/broker_test.go:
```diff
@@ -208,12 +229,82 @@ func TestValidateServiceBroker(t *testing.T) {
+				},
+			},
+			valid: false,
+		},
```
need test for negative relistRequests

nilebox commented on pkg/apis/servicecatalog/v1alpha1/defaults.go:
```diff
@@ -18,13 +18,28 @@ package v1alpha1
 }
 
+func SetDefaults_ServiceBrokerSpec(spec *ServiceBrokerSpec) {
+	isValidRelistBehavior := spec.RelistBehavior == "Duration" ||
```
Reference `ServiceBrokerRelistBehaviorDuration` and `ServiceBrokerRelistBehaviorManual` constants instead of inline strings?

nilebox replied:
This code shouldn't be there at all, see below.

nilebox commented on pkg/apis/servicecatalog/v1alpha1/defaults.go:
```diff
@@ -18,13 +18,28 @@ package v1alpha1
+	isValidRelistBehavior := spec.RelistBehavior == "Duration" ||
+		spec.RelistBehavior == "Manual"
+
+	if !isValidRelistBehavior {
```
Defaults shouldn't handle invalid values, they handle missing values.
Hence, instead of the check above you should check for empty string here.

nilebox commented on pkg/apis/servicecatalog/validation/broker.go:
```diff
@@ -98,13 +99,50 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
+		)
+	}
+
+	if spec.RelistBehavior == sc.ServiceBrokerRelistBehaviorDuration && spec.RelistDuration == nil {
```
as mentioned in the other comments thread, you should also check that `RelistDuration` is nil when `RelistBehavior != Duration`

nilebox commented on pkg/apis/servicecatalog/validation/broker_test.go:
```diff
@@ -40,7 +41,9 @@ func TestValidateServiceBroker(t *testing.T) {
 				Spec: servicecatalog.ServiceBrokerSpec{
-					URL: "http://example.com",
+					URL:            "http://example.com",
+					RelistBehavior: "Duration",
```
use constant from types (here and below).

nilebox commented on pkg/apis/servicecatalog/validation/broker_test.go:
```diff
@@ -40,7 +41,9 @@ func TestValidateServiceBroker(t *testing.T) {
-					URL: "http://example.com",
+					URL:            "http://example.com",
+					RelistBehavior: "Duration",
+					RelistDuration: &metav1.Duration{15 & time.Minute},
```
should be `*` instead of `&`. here and below.

nilebox commented on pkg/storage/tpr/storage_interface_test.go:
```diff
@@ -140,7 +141,9 @@ func TestCreateWithNoNamespace(t *testing.T) {
 		Spec: sc.ServiceBrokerSpec{
-			URL: "http://my-awesome-broker.io",
+			URL:            "http://my-awesome-broker.io",
+			RelistBehavior: "Duration",
```
use constant

nilebox commented on pkg/storage/tpr/storage_interface_test.go:
```diff
@@ -140,7 +141,9 @@ func TestCreateWithNoNamespace(t *testing.T) {
-			URL: "http://my-awesome-broker.io",
+			URL:            "http://my-awesome-broker.io",
+			RelistBehavior: "Duration",
+			RelistDuration: &metav1.Duration{15 * time.Minute},
```
What are we trying to check there by setting those fields? If we want to make sure that serialization and persistence works correctly, you should set values different from default.

eriknelson commented:
@nilebox @pmorie think I should have addressed your comments in the last couple commits, also rebased and repaired conflicts. Let me know if I missed something, thank you for the reviews!

nilebox commented on pkg/controller/controller_broker_test.go:
```diff
@@ -133,12 +138,33 @@ func TestShouldReconcileServiceBroker(t *testing.T) {
+			name: "ready, duration behavior, nil duration",
+			broker: func() *v1alpha1.ServiceBroker {
+				broker := getTestServiceBrokerWithStatus(v1alpha1.ConditionTrue)
+				broker.Spec.RelistBehavior = "Duration"
```
const

nilebox commented on pkg/apis/servicecatalog/testing/fuzzer.go:
```diff
@@ -192,6 +193,11 @@ func FuzzerFor(t *testing.T, version schema.GroupVersion, src rand.Source) *fuzz
 		},
+		func(bs *servicecatalog.ServiceBrokerSpec, c fuzz.Continue) {
+			c.FuzzNoCustom(bs)
+			bs.RelistBehavior = "Duration"
```
const

nilebox commented on pkg/controller/controller_broker_test.go:
```diff
@@ -133,12 +138,33 @@ func TestShouldReconcileServiceBroker(t *testing.T) {
+			name: "ready, manual behavior",
+			broker: func() *v1alpha1.ServiceBroker {
+				broker := getTestServiceBrokerWithStatus(v1alpha1.ConditionTrue)
+				broker.Spec.RelistBehavior = "Manual"
```
const

eriknelson commented:
@nilebox think I got the last of them. For my edification, I had someone suggest it might be preferred to not use the consts from types in tests in case the values change? I'm probably misunderstanding that, but are there any cases where an inline value might be preferred?

nilebox commented:
@eriknelson if you want to ensure that the constant hasn't changed, I would write a specific test for it. Might be something as trivial as
```
if RelistDuration != "Duration" {
  t.Fail();
}
```

pmorie commented on pkg/apis/servicecatalog/validation/broker.go:
```diff
@@ -98,13 +98,49 @@ func validateServiceBrokerSpec(spec *sc.ServiceBrokerSpec, fldPath *field.Path)
+	if spec.RelistBehavior == sc.ServiceBrokerRelistBehaviorManual && spec.RelistDuration != nil {
+		allErrs = append(
+			allErrs,
+			field.Required(fldPath.Child("relistDuration"), "relistDuration must not be set if behavior is set to Manual"),
```
s/behavior/relistBehavior/

pmorie commented:
/test

pmorie commented:
LGTM
